File: DataStructures/Heap/heap.ts
// BinaryHeap

type f = (n: number) => boolean;
class Heap extends Array {
  #hasParent: f = (index) => index > 0;

  #hasLeftChild(index: number) {
    return index * 2 <= this.length;
  }
  #hasRightChild(index: number) {
    return index * 2 + 1 <= this.length;
  }
  #getParentIndex(index: number) {
    return Math.floor((index - 1) / 2);
  }
  #getLeftChildIndex(index: number) {
    return index * 2 + 1;
  }
  #getRightChildIndex(index: number) {
    return index * 2 + 2;
  }
  #getParent(index: number) {
    return this[this.#getParentIndex(index)];
  }
  #getLeftChild(index: number) {
    return this[this.#getLeftChildIndex(index)];
  }
  #getRightChild(index: number) {
    return this[this.#getRightChildIndex(index)];
  }

  add(d: number) {
    this.push(d);
    this.heapifyUp(this.length - 1);
  }
  heapifyUp(index: number): void {
    if (!this.#hasParent(index)) return;
    if (this[index] < this.#getParent(index)) {
      let pindex = this.#getParentIndex(index);
      this.swap(index, pindex);
      this.heapifyUp(pindex);
    }
  }
  swap(indexA: number, indexB: number) {
    if (
      indexA < 0 ||
      indexA > this.length - 1 ||
      indexB < 0 ||
      indexB > this.length - 1
    )
      return;
    let temp = this[indexA];
    this[indexA] = this[indexB];
    this[indexB] = temp;
  }
  // hasParent()
  //  getParent(index)
  //      getParentIndex(index)
  //

  poll() {
    if (this.length === 0) return;
    let min = this[0];
    this[0] = this.pop();
    this.heapifyDown(0);
    return min;
  }
  heapifyDown(index: number) {
    if (this.#hasLeftChild(index)) {
      //child index
      let ci;
      if (this.#hasRightChild(index)) {
        ci =
          this.#getLeftChild(index) <= this.#getRightChild(index)
            ? this.#getLeftChildIndex(index)
            : this.#getRightChildIndex(index);
      } else ci = this.#getLeftChildIndex(index);
      this.swap(index, ci);
      this.heapifyDown(ci);
    }
  }

  // sort
  peek(): number {
    return this[0];
  }
}

let h = new Heap()
h.add(50)
h.add(20)
h.add(30)
h.add(10)
h.add(5)
h.add(7)
h.add(90)
console.log(h)

h.poll()
console.log(h)
h.poll()
console.log(h)

File: DataStructures/Heap/heap.py

class MinHeap:
    'there is no gaps in a heap, therefor implementing it as an array'
    def __init__(self):
        self.size = 0
        self.array = []

    def get_left_child_index(parent_index): return parent_index * 2
    def get_right_child_index(parent_index): return parent_index * 2 + 1
    def get_parent_index(child_index): return child_index // 2

    def has_left_child(self, index): return get_left_child_index(index) < self.size
    def has_right_child(self, index): return get_left_child_index(index) < self.size
    def has_parent(self, index): return get_parent_index(index) >= 0

    def left_child(self, index): return self.array[get_left_child_index(index)]
    def right_child(self, index): return self.array[get_right_child_index(index)]
    def parent(self, index): return self.array[get_parent_index(index)]

    def swap(self, index1, index2):
        self.array[index1]. self.array[index2] = self.array[index2], self.array[index1]
    
    def peek(self):
        if self.size == 0: return None
        return self.array[0]
    
    def poll(self):
        if self.size == 0: return None
        item = self.array[0]
        self.array[0] = self.array[-1]
        self.size -=1
        self.heapify_down()
        return item

    def add(self, item):
        self.array.append(item)
        self.size +=1
        self.heapify_up()

    def heapify_down(self):
        index = 0
        while self.has_left_child(index):
            if (self.has_right_child(index) and self.right_child(index) < self.left_child(index)):
                index = self.get_right_child_index(index)
            else:
                index = self.get_left_child_index(index)
            if  self.array[index] >= self.parent(index):
                break
            else:
                self.swap(index, self.get_parent_index(index))
        
    def heapify_up(self):
        index = self.size - 1
        while self.has_parent(index) and self.array[index] < self.parent(index):
            parentIndex = self.get_parent_index(index)
            self.swap(index,parentIndex) 
            index = self.parentIndex